client/starwhale/core/job/model.py

```diff
@@ -60,10 +95,89 @@ def execute(self) -> bool:
             func(context=self.context)
 
         except Exception as e:
+            self.exception = e
             self.status = STATUS.FAILED
-            logger.error(f"execute step:{self.context}, error:{e}")
-            return False
         else:
             self.status = STATUS.SUCCESS
-            logger.debug(f"execute step:{self.context} success")
-            return True
+        finally:
+            logger.debug(
+                f"execute step:{self.context}, status:{self.status}, error:{self.exception}"
+            )
+            return TaskResult(
+                task_id=self.index, status=self.status, exception=self.exception
+            )
+
+
+class StepExecutor:
+    def __init__(
+        self,
+        step: Step,
+        project: str,
+        version: str,
+        module: str,
+        workdir: Path,
+        dataset_uris: t.List[str],
+        kw: t.Dict[str, t.Any] = {},
+    ) -> None:
+        self.step = step
+        self.project = project
+        self.dataset_uris = dataset_uris
+        self.module = module
+        self.workdir = workdir
+        self.version = version
+        self.kw = kw
+
+    def execute(self) -> StepResult:
+        logger.debug(f"start execute step:{self.step}")
+        processor = MultiThreadProcessor(
+            self.step.step_name, self.step.concurrency, self._split_tasks()
+        )
+        task_results = processor.exec()
+        logger.debug(f"finish execute step:{self.step}")
+        return StepResult(step_name=self.step.step_name, task_results=task_results)
+
+    def _split_tasks(self) -> t.List[BaseExecutor]:
+        return list(
+            TaskExecutor(
+                index=index,
+                context=Context(
+                    project=self.project,
+                    version=self.version,
+                    step=self.step.step_name,
+                    total=self.step.task_num,
+                    index=index,
+                    dataset_uris=self.dataset_uris,
+                    workdir=self.workdir,
+                    kw=self.kw,
+                ),
+                status=STATUS.INIT,
+                module=self.module,
+                workdir=self.workdir,
+            )
+            for index in range(self.step.task_num)
+        )
+
+
+class MultiThreadProcessor:
+    def __init__(
+        self,
+        name: str,
+        concurrency: int,
+        executors: t.List[BaseExecutor],
+    ) -> None:
+        self.name = name
+        self.concurrency = concurrency
+        self.executors = executors
+
+    def exec(self) -> t.List[t.Any]:
+        with concurrent.futures.ThreadPoolExecutor(
+            max_workers=self.concurrency
+        ) as pool:
+            start_time = time.time()
+            futures = [pool.submit(executor.execute) for executor in self.executors]
+            _results: t.List[t.Any] = list(
+                future.result() for future in concurrent.futures.as_completed(futures)
+            )
```

Comment:
_results: t.List[t.Any] = list(
                future.result() for future in concurrent.futures.as_completed(futures)
            )
-->
_results: t.List[t.Any] = [future.result() for future in concurrent.futures.as_completed(futures)]